src/Middleware/RequestThrottling/test/MiddlewareTests.cs

```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Aspnetcore.RequestThrottling;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Microsoft.AspNetCore.RequestThrottling.Tests
+{
+    public class MiddlewareTests
+    {
+        [Fact]
+        public async Task RequestsCanEnterIfSpaceAvailible()
+        {
+            var middleware = TestUtils.CreateTestMiddleWare(maxConcurrentRequests: 1);
+            var context = new DefaultHttpContext();
+
+            // a request should go through with no problems
+            await middleware.Invoke(context).OrTimeout();
+        }
+
+        [Fact]
+        public async Task RequestsAreBlockedIfNoSpaceAvailible()
+        {
+            var blocker = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var firstRequest = true;
+
+            var middleware = TestUtils.CreateTestMiddleWare(
+                maxConcurrentRequests: 1,
+                next: httpContext =>
+                {
+                    if (firstRequest)
+                    {
+                        firstRequest = false;
```

Comment:
You can use another TaskCompletionSource to ensure the first request was received before sending the second. You can await it after getting t1 below, and complete it here after setting firstRequest.

If you had enough async operations going on at once you wouldn't be able to guarantee that t1 and t2 didn't get re-ordered. I don't expect such in a simple case like this but it's better to have a positive signal from the app that the first request was received before sending the second.

Reply:
This looks like a job for... [`SyncPoint`](https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/common/testassets/Tests.Utils/SyncPoint.cs)!

![An animated superman flying through the sky](https://media.giphy.com/media/10bKPDUM5H7m7u/giphy.gif)

Reply:
For more complicated tests, `SyncPoint` will be useful. This test is okay, because the code that retrieves the tasks for t1 and t2 runs synchronously, so it can't get out of order.

Reply:
@anurse I showed @DylanDmitri SyncPoint yesterday because that's what I immediately though of too. Useful type there 😉